### ISSUE TITLE
Example usage of tight/sending JPEG data?

### ISSUE THREAD

**Issue by george-geng** [LibVNC/libvncserver#321]
Hi, I was wondering if there are any examples (or documentation) which show how to use the functionality in `tight.c`, specifically, sending JPG data to a client.

I saw that this similar [question ](https://github.com/LibVNC/libvncserver/issues/232)and that this [PR](https://github.com/LibVNC/libvncserver/pull/235) should make it possible to do so, but I'm not sure how to go about implementing this in my own vcn server code, since the examples I'm looking at (such as `pnmshow.c`) all just seem to fill the frameBuffer instead.

Any example code or pointers in the right direction would be super appreciated, thanks!

**Comment by bk138** [comment]
Hi! `tight.c` does tight encoding of whatever raw image data you feed to it. Do you want the streamed data to be JPEG-compressed (that's built-in) or is your source data JPEG?

**Comment by george-geng** [comment]
Hello! My source data would a raw RGBA raster, that I would like to stream as JPEG-compressed data.

By that's built-in, do you mean that the server example that reads in an PNM image and displays it is already sending JPEG compressed data over to a vnc client? (And likewise, the client example `ppmtest.c` already performs JPEG decompression?)

**Comment by bk138** [comment]
Yes, indeed. Please look at the log output:

```
./pnmshow a.pnm 
04/07/2019 11:19:06 Got width 512 and height 768.
04/07/2019 11:19:06 Listening for VNC connections on TCP port 5900
04/07/2019 11:19:06 Listening for VNC connections on TCP6 port 5900
04/07/2019 11:19:06 Listening for HTTP connections on TCP port 5800
04/07/2019 11:19:06   URL http://denkpad:5800
04/07/2019 11:19:06 Listening for HTTP connections on TCP6 port 5800
04/07/2019 11:19:06   URL http://denkpad:5800
04/07/2019 11:19:11 Got connection from client 127.0.0.1
04/07/2019 11:19:11   0 other clients
04/07/2019 11:19:11 Normal socket connection
04/07/2019 11:19:11 Client Protocol Version 3.8
04/07/2019 11:19:11 Protocol version sent 3.8, using 3.8
04/07/2019 11:19:11 rfbProcessClientSecurityType: executing handler for type 1
04/07/2019 11:19:11 rfbProcessClientSecurityType: returning securityResult for client rfb version >= 3.8
04/07/2019 11:19:11 Pixel format for client 127.0.0.1:
04/07/2019 11:19:11   32 bpp, depth 24, little endian
04/07/2019 11:19:11   true colour: max r 255 g 255 b 255, shift r 16 g 8 b 0
04/07/2019 11:19:11 rfbProcessClientNormalMessage: ignoring unsupported encoding type ultraZip
04/07/2019 11:19:11 Using compression level 3 for client 127.0.0.1
04/07/2019 11:19:11 Using image quality level 5 for client 127.0.0.1
04/07/2019 11:19:11 Using JPEG subsampling 2, Q77 for client 127.0.0.1
04/07/2019 11:19:11 Enabling KeyboardLedState protocol extension for client 127.0.0.1
04/07/2019 11:19:11 Enabling NewFBSize protocol extension for client 127.0.0.1
04/07/2019 11:19:11 Enabling LastRect protocol extension for client 127.0.0.1
04/07/2019 11:19:11 Enabling SupportedMessages protocol extension for client 127.0.0.1
04/07/2019 11:19:11 Enabling SupportedEncodings protocol extension for client 127.0.0.1
04/07/2019 11:19:11 Enabling ServerIdentity protocol extension for client 127.0.0.1
04/07/2019 11:19:11 Using tight encoding for client 127.0.0.1
```

**Comment by george-geng** [comment]
Ah, thank you. that looks great! This was my output log:

```
04/07/2019 12:38:17 Got width 1920 and height 1080.
04/07/2019 12:38:17 Listening for VNC connections on TCP port 5900
04/07/2019 12:38:17 Listening for VNC connections on TCP6 port 5900
04/07/2019 12:38:17 Listening for HTTP connections on TCP port 5800
04/07/2019 12:38:17   URL http://LAPTOP-3I2G558F:5800
04/07/2019 12:38:17 Listening for HTTP connections on TCP6 port 5800
04/07/2019 12:38:17   URL http://LAPTOP-3I2G558F:5800
04/07/2019 12:38:46 Got connection from client 127.0.0.1
04/07/2019 12:38:46   other clients:
04/07/2019 12:38:46 Normal socket connection
04/07/2019 12:38:46 Client Protocol Version 3.8
04/07/2019 12:38:46 Protocol version sent 3.8, using 3.8
04/07/2019 12:38:46 rfbProcessClientSecurityType: executing handler for type 1
04/07/2019 12:38:46 rfbProcessClientSecurityType: returning securityResult for client rfb version >= 3.8
04/07/2019 12:38:46 Pixel format for client 127.0.0.1:
04/07/2019 12:38:46   32 bpp, depth 24, little endian
04/07/2019 12:38:46   true colour: max r 255 g 255 b 255, shift r 0 g 8 b 16
04/07/2019 12:38:46 rfbProcessClientNormalMessage: ignoring unsupported encoding type ultraZip
04/07/2019 12:38:46 Enabling KeyboardLedState protocol extension for client 127.0.0.1
04/07/2019 12:38:46 Enabling SupportedMessages protocol extension for client 127.0.0.1
04/07/2019 12:38:46 Enabling SupportedEncodings protocol extension for client 127.0.0.1
04/07/2019 12:38:46 Enabling ServerIdentity protocol extension for client 127.0.0.1
04/07/2019 12:38:46 Enabling Xvp protocol extension for client 127.0.0.1
04/07/2019 12:38:46 Using ultra encoding for client 127.0.0.1
```

which is why I was wondering about how to get it to perform tight encoding and JPEG compression. I'm thinking that this is an issue with my build settings then, "For the more advanced VNC encodings: zlib and libjpeg development packages." If that's the case I will try re-buildling with zlib and libjpeg.

Just 2 more questions before I think this thread can be closed.

1. I noticed this: 
```04/07/2019 11:19:11 Using compression level 3 for client 127.0.0.1
04/07/2019 11:19:11 Using image quality level 5 for client 127.0.0.1```
Is there a way to specify the compression/image quality level in code? I couldn't see anything suggestive of that inside of `ppmtest.c`.

2. In order for me to get `ppmtest.c` to save out a complete image (send out by `pnmshow.c`), I had to comment out these lines inside of `SaveFramebufferAsPPM`
```
    /* save one picture only if the last is older than 2 seconds */
    t1 = time(NULL);
    if (t1 - t>2)
        t = t1;
    else
        return;
```

Am I just sending too big of an image (1920 x 1080)? Or is there something else I'm misunderstanding here?

Many thanks!

**Comment by bk138** [comment]
Please read https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst on how VNC/RFB works, this might clear up some things :-)

tl;dr: the client requests the compresslevel and quality. Please grep the source to see what you have to set. 

If you'd like to improve the examples with the bits setting compress and quality level, a PR is most appreciated!

**Comment by george-geng** [comment]
sounds good, thanks for the link and clarifications!
